In [61]:
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import numpy as np
from scipy import signal
import scipy.io
from scipy import stats

## Questão 4

## Item a -

Funções para os cálculos dos parâmetros listados.

In [13]:
def RMS(x):
    rms = np.sqrt((np.sum(np.abs(np.power(x,2)))/len(x)))
    return rms

def VMR(x):
    vmr = np.sqrt((np.sum(np.abs(x)))/len(x))
    return vmr

def mean_frequency(x,fs):
    X = np.power(np.abs(np.fft.fft(x)),2)
    X = X[0:int(len(x)/2)]
    pesos = np.arange(0,len(X),1)
    kmed = np.sum(np.multiply(X,pesos))/np.sum(X)
    fmed = kmed * fs/ len(x)
    return fmed

def modal_frequency(x,fs):
    X = np.power(np.abs(np.fft.fft(x)),2)
    X = X[0:int(len(x)/2)]
    kmod = np.argmax(X)
    fmod = kmod * fs/ len(x)
    return fmod

def median_frequency(x,fs):
    X = np.power(np.abs(np.fft.fft(x)),2)
    X = X[0:int(len(x)/2)]
    total_energy = np.sum(X)
    
    i = 0
    
    while True:
        sinal_teste = X[0:i]
        energy_calculated = np.sum(np.abs(np.power(sinal_teste,2)))
        if (energy_calculated >= total_energy/2):
            
            break
        else:
            i += 1
          
    fmedian = i * fs / len(x)    
    return fmedian

    


In [14]:
def parametros(sinal,janela,passo,fs,win = None):
    sinal = sinal - np.mean(sinal)
    janela = int(fs * janela) 
    passo = int(fs * passo)
    
    #ESCOLHA DA JANELA
    
    if (win == None):
        win = np.hamming(janela)
    elif (win == 'Hamming' or win == 'hamming'):
        win = np.hamming(janela)
    elif (win == 'Hanning' or win == 'hanning'):
        win = np.hanning(janela)
    elif (win == 'Blackman' or win == 'blackman'):
        win = np.blackman(janela) 
    elif (win == 'Bartlett' or win == 'bartlett'):
        win = np.bartlett(janela)     
    else: 
        return print('Erro: Janela não especificada')
    
    #INICIALIZAÇÃO DAS VARIÁVEIS
    
    init = 0
    RMS_ = []
    VMR_ = []
    FMED_ = []
    FMOD_ = []
    FMEDIAN_ = []
    
    #SEPARAÇÃO EM JANELAS
    
    while True:   
        amostras = sinal[init:((init + janela))]
        if(len(amostras) == janela):
            
            #CALCULOS DOS PARAMETROS
            
            rms = RMS(amostras)
            RMS_ = np.append(RMS_,rms)
            
            vmr = VMR(amostras)
            VMR_ = np.append(VMR_,vmr)
            
            #MULTIPLICAÇÃO PELA JANELA
            
            amostras_janeladas = np.multiply(amostras,win)
            
            fmed = mean_frequency(amostras_janeladas,fs)
            FMED_ = np.append(FMED_,fmed)
            
            fmod = modal_frequency(amostras_janeladas,fs)
            FMOD_ = np.append(FMOD_,fmod)
            
            fmedian = median_frequency(amostras_janeladas,fs)
            FMEDIAN_ = np.append(FMEDIAN_,fmedian)
            
            init += passo
            
        else:
            break
    
    return RMS_,VMR_,FMED_,FMOD_,FMEDIAN_

## Carregando sinais de cada grupo

In [158]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/a')
data = scipy.io.loadmat('x1.mat')
y = data['x']
xa = [row[0] for row in y]
fs_mat = data['fs']
fsa = fs_mat[0][0]

In [159]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/b')
data1 = scipy.io.loadmat('x101.mat')
y1 = data1['x']
xb = [row[0] for row in y1]
fs_mat1 = data1['fs']
fsb = fs_mat1[0][0]

In [160]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/c')
data2 = scipy.io.loadmat('x201.mat')
y2 = data2['x']
xc = [row[0] for row in y2]
fs_mat2 = data2['fs']
fsc = fs_mat2[0][0]

In [165]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/d')
data3 = scipy.io.loadmat('x301.mat')
y3 = data3['x']
xd = [row[0] for row in y3]
fs_mat3 = data3['fs']
fsd = fs_mat3[0][0]

In [169]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/e')
data4 = scipy.io.loadmat('x401.mat')
y4 = data4['x']
xe = [row[0] for row in y4]
fs_mat4 = data4['fs']
fse = fs_mat4[0][0]

## Calculando RMS,VMR,FMED,FMOD e FMEDIAN para os sinais

## Grupo a

In [204]:
RMS_a,VMR_a,FMED_a,FMOD_a,FMEDIAN_a = parametros(xa,5,2.5,fsa,win = None)
RMS_a = np.round(np.mean(RMS_a),2)
VMR_a = np.round(np.mean(VMR_a),2)
FMED_a = np.round(np.mean(FMED_a),2)
FMOD_a = np.round(np.mean(FMOD_a),2)
FMEDIAN_a = np.round(np.mean(FMEDIAN_a),2)
print("RMS = "+ str(RMS_a) + " VMR = " + str(VMR_a) + " FMED = " + str(FMED_a)\
+ " FMOD = " + str(FMOD_a) + " FMEDIAN = " + str(FMEDIAN_a) + " (*valores médios)")

RMS = 28.72 VMR = 4.84 FMED = 3.55 FMOD = 1.2 FMEDIAN = 0.2 (*valores médios)


## Grupo b

In [196]:
RMS_b,VMR_b,FMED_b,FMOD_b,FMEDIAN_b = parametros(xb,5,2.5,fsb,win = None)
RMS_b = np.round(np.mean(RMS_b),2)
VMR_b = np.round(np.mean(VMR_b),2)
FMED_b = np.round(np.mean(FMED_b),2)
FMOD_b = np.round(np.mean(FMOD_b))
FMEDIAN_b = np.round(np.mean(FMEDIAN_b),2)
print("RMS = "+ str(RMS_b) + " VMR = " + str(VMR_b) + " FMED = " + str(FMED_b)\
+ " FMOD = " + str(FMOD_b) + " FMEDIAN = " + str(FMEDIAN_b) + " (*valores médios)")

RMS = 48.42 VMR = 6.18 FMED = 3.66 FMOD = 2.0 FMEDIAN = 0.23 (*valores médios)


## Grupo c

In [197]:
RMS_c,VMR_c,FMED_c,FMOD_c,FMEDIAN_c = parametros(xc,5,2.5,fsc,win = None)
RMS_c = np.round(np.mean(RMS_c),2)
VMR_c = np.round(np.mean(VMR_c),2)
FMED_c = np.round(np.mean(FMED_c),2)
FMOD_c = np.round(np.mean(FMOD_c))
FMEDIAN_c = np.round(np.mean(FMEDIAN_c),2)
print("RMS = "+ str(RMS_c) + " VMR = " + str(VMR_c) + " FMED = " + str(FMED_c)\
+ " FMOD = " + str(FMOD_c) + " FMEDIAN = " + str(FMEDIAN_c) + " (*valores médios)")

RMS = 49.05 VMR = 6.25 FMED = 7.41 FMOD = 2.0 FMEDIAN = 0.23 (*valores médios)


## Grupo d

In [203]:
RMS_d,VMR_d,FMED_d,FMOD_d,FMEDIAN_d = parametros(xd,5,2.5,fsd,win = None)
RMS_d = np.round(np.mean(RMS_d),2)
VMR_d = np.round(np.mean(VMR_d),2)
FMED_d = np.round(np.mean(FMED_d),2)
FMOD_d = np.round(np.mean(FMOD_d),2)
FMEDIAN_d = np.round(np.mean(FMEDIAN_d),2)
print("RMS = "+ str(RMS_d) + " VMR = " + str(VMR_d) + " FMED = " + str(FMED_d)\
+ " FMOD = " + str(FMOD_d) + " FMEDIAN = " + str(FMEDIAN_d) + " (*valores médios)")

RMS = 474.38 VMR = 18.97 FMED = 8.92 FMOD = 3.48 FMEDIAN = 0.2 (*valores médios)


## Grupo e

In [206]:
RMS_e,VMR_e,FMED_e,FMOD_e,FMEDIAN_e = parametros(xe,5,2.5,fse,win = None)
RMS_e = np.round(np.mean(RMS_e),2)
VMR_e = np.round(np.mean(VMR_e),2)
FMED_e = np.round(np.mean(FMED_e),2)
FMOD_e = np.round(np.mean(FMOD_e),2)
FMEDIAN_e = np.round(np.mean(FMEDIAN_e),2)
print("RMS = "+ str(RMS_e) + " VMR = " + str(VMR_e) + " FMED = " + str(FMED_e)\
+ " FMOD = " + str(FMOD_e) + " FMEDIAN = " + str(FMEDIAN_e) + " (*valores médios)")

RMS = 42.87 VMR = 5.79 FMED = 6.83 FMOD = 2.25 FMEDIAN = 0.2 (*valores médios)


## Tabela de Resultados

| GRUPO         | A     | B     | C     |    D   | E     |
|---------------|-------|-------|-------|:------:|-------|
| RMS (uM)      | 28.72 | 48.42 | 49.05 | 474.38 | 42.87 |
| VMR (uM)      | 4.84  | 6.18  | 6.25  | 18.97  | 5.79  |
| FMEDIA (Hz)   | 3.55  | 3.66  | 7.41  | 8.92   | 6.83  |
| FMODAL (Hz)   | 1.2   | 2.0   | 2.0   | 3.48   | 2.25  |
| FMEDIANA (Hz) | 0.2   | 0.23  | 0.23  | 0.2    | 0.2   |

## Item b -

## Calculando energia nas bandas delta, teta, alfa, beta, gama

In [21]:
def energia_bandas(sinal,fs):
    
    #Bandas
    delta = 4
    teta = [4,7]
    alfa = [7,15]
    beta = [16,31]
    gama = 32
    
    #filtros FIR nas frequências da tabela do item b
    
    Wdelta = signal.firwin(1001, delta, nyq=fs/2)
    Wteta = signal.firwin(1001, teta, nyq=fs/2, pass_zero=False)
    Walfa = signal.firwin(1001, alfa, nyq=fs/2, pass_zero=False)
    Wbeta = signal.firwin(1001, beta, nyq=fs/2, pass_zero=False)
    Wgama = signal.firwin(1001, gama, nyq=fs/2, pass_zero=False)
    
    sinal_delta = np.convolve(sinal,Wdelta)
    sinal_teta = np.convolve(sinal,Wteta)
    sinal_alfa = np.convolve(sinal,Walfa)
    sinal_beta = np.convolve(sinal,Wbeta)
    sinal_gama = np.convolve(sinal,Wgama)
    
    #calculando energia nas bandas
    
    energia_total = np.sum(np.power(np.abs((sinal)),2))
    
    energia_delta = np.sum(np.power(np.abs(sinal_delta),2))
    energia_teta = np.sum(np.power(np.abs(sinal_teta),2))
    energia_alfa = np.sum(np.power(np.abs(sinal_alfa),2))
    energia_beta = np.sum(np.power(np.abs(sinal_beta),2))
    energia_gama = np.sum(np.power(np.abs(sinal_gama),2))
    
    energia_percentual_delta = energia_delta/energia_total * 100
    energia_percentual_teta = energia_teta/energia_total * 100
    energia_percentual_alfa = energia_alfa/energia_total * 100
    energia_percentual_beta = energia_beta/energia_total * 100
    energia_percentual_gama = energia_gama/energia_total * 100
    
    #print(energia_total, energia_delta)
    
    return energia_percentual_delta,energia_percentual_teta,energia_percentual_alfa,\
    energia_percentual_beta,energia_percentual_gama
    
def calcula_energia(sinal,fs,janela,passo):
    sinal = sinal - np.mean(sinal)
    janela = int(fs * janela) 
    passo = int(fs * passo)
     
    init = 0
    AMOSTRAS = []
    DELTA = []
    TETA = []
    ALFA = []
    BETA = []
    GAMA = []
    
    while True:   
        amostras = sinal[init:(init + janela)]
        if(len(amostras) == janela):
            
            delta,teta,alfa,beta,gama = energia_bandas(amostras,fs)
            
            
            
            
            DELTA = np.append(DELTA,delta)
            TETA = np.append(TETA,teta)
            ALFA = np.append(ALFA,alfa)
            BETA = np.append(BETA,beta)
            GAMA = np.append(GAMA,gama)
            
            init += passo
    
        else:
            break
    
    return DELTA,TETA,ALFA,BETA,GAMA
    

## Cálculo de energia nas bandas para os sinais dos grupos A, B, C, D e E

## Grupo A

In [315]:
q,w,e,r,s = calcula_energia(xa,fsa,5,2.5)

In [319]:
print(q,w,e,r,s)

75.03875661797926 11.710278896902995 8.384495602810858 2.168638904657522 0.5516004338818751


## Grupo B

In [321]:
q1,w1,e1,r1,s1 = calcula_energia(xb,fsb,5,2.5)

In [322]:
print(q1,w1,e1,r1,s1)

67.1403345904578 19.935929427275067 9.632101803392633 1.0958794908869482 0.19349837786758772


## Grupo C

In [324]:
q2,w2,e2,r2,s2 = calcula_energia(xc,fsc,5,2.5)

In [325]:
print(q2,w2,e2,r2,s2)

41.5739883241532 9.129799457146959 40.35035308423433 5.9692827275536775 0.7438470536058548


## Grupo D

In [327]:
q3,w3,e3,r3,s3 = calcula_energia(xd,fsd,5,2.5)

In [328]:
print(q3,w3,e3,r3,s3)

31.251221054645605 16.125306367217224 31.515451675540643 13.69676440449765 0.41606828730334766


## Grupo E

In [329]:
q4,w4,e4,r4,s4 = calcula_energia(xe,fse,5,2.5)

In [330]:
print(q4,w4,e4,r4,s4)

41.27014174892743 15.13906173557773 32.47235942619969 7.4534937067700096 0.5673841511483435


## Resultados (**valores percentuais)

| BANDAS\GRUPO |   A   |   B   |   C   |   D   |   E   |
|--------------|:-----:|:-----:|:-----:|:-----:|:-----:|
|     DELTA    | 75.04 | 67.14 | 41.57 | 31.25 | 41.27 |
|     TETA     | 11.71 | 19.94 | 9.13  | 16.13 | 15.14 |
|     ALFA     | 8.38  | 9.63  | 40.35 | 31.52 | 32.47 |
|     BETA     | 2.17  | 1.01  | 5.97  | 13.70 | 7.45  |
|     GAMA     | 0.55  | 0.19  | 0.74  | 0.42  | 0.57  |

## Item c -

In [78]:
def energias_em_bandas_consecutivas(sinal,fs):
    
    nbandas = 10
    
    max_f = fs/2
    
    band_increment = max_f/nbandas
    
    bandas = []
    
    for i in range(1,nbandas):
        bandas.append(band_increment * i)
        
    n1,d1 = signal.butter(5, bandas[0]/fs * 2, 'low')
    n2,d2 = signal.butter(5, [bandas[0]/fs,bandas[1]/fs]* 2, 'bandpass' )
    n3,d3 = signal.butter(5, [bandas[1]/fs,bandas[2]/fs]* 2, 'bandpass')
    n4,d4 = signal.butter(5, [bandas[2]/fs,bandas[3]/fs]* 2, 'bandpass')
    n5,d5 = signal.butter(5, [bandas[3]/fs,bandas[4]/fs]* 2, 'bandpass')
    n6,d6 = signal.butter(5, [bandas[4]/fs,bandas[5]/fs]* 2, 'bandpass')
    n7,d7 = signal.butter(5, [bandas[5]/fs,bandas[6]/fs]* 2, 'bandpass')
    n8,d8 = signal.butter(5, [bandas[6]/fs,bandas[7]/fs]* 2, 'bandpass')
    n9,d9 = signal.butter(5, [bandas[7]/fs,bandas[8]/fs]* 2, 'bandpass')
    n10,d10 = signal.butter(5, bandas[8]/fs, 'highpass')
    
    sinal1 = scipy.signal.lfilter(n1, d1, sinal)
    sinal2 = scipy.signal.lfilter(n2, d2, sinal)
    sinal3 = scipy.signal.lfilter(n3, d3, sinal)
    sinal4 = scipy.signal.lfilter(n4, d4, sinal)
    sinal5 = scipy.signal.lfilter(n5, d5, sinal)
    sinal6 = scipy.signal.lfilter(n6, d6, sinal)
    sinal7 = scipy.signal.lfilter(n7, d7, sinal)
    sinal8 = scipy.signal.lfilter(n8, d8, sinal)
    sinal9 = scipy.signal.lfilter(n9, d9, sinal)
    sinal10 = scipy.signal.lfilter(n10, d10, sinal)
    
    energia_total = np.sum(np.power(np.abs(sinal),2))
    energia_1 = np.sum(np.power(np.abs(sinal1),2))
    energia_2 = np.sum(np.power(np.abs(sinal2),2))
    energia_3 = np.sum(np.power(np.abs(sinal3),2))
    energia_4 = np.sum(np.power(np.abs(sinal4),2))
    energia_5 = np.sum(np.power(np.abs(sinal5),2))
    energia_6 = np.sum(np.power(np.abs(sinal6),2))
    energia_7 = np.sum(np.power(np.abs(sinal7),2))
    energia_8 = np.sum(np.power(np.abs(sinal8),2))
    energia_9 = np.sum(np.power(np.abs(sinal9),2))
    energia_10 = np.sum(np.power(np.abs(sinal10),2))
    
    energia_percentual1 = energia_1/energia_total * 100
    energia_percentual2 = energia_2/energia_total * 100
    energia_percentual3 = energia_3/energia_total * 100
    energia_percentual4 = energia_4/energia_total * 100
    energia_percentual5 = energia_5/energia_total * 100
    energia_percentual6 = energia_6/energia_total * 100
    energia_percentual7 = energia_7/energia_total * 100
    energia_percentual8 = energia_8/energia_total * 100
    energia_percentual9 = energia_9/energia_total * 100
    energia_percentual10 = energia_10/energia_total * 100
        
    return (energia_percentual1,energia_percentual2,energia_percentual3,energia_percentual4,energia_percentual5\
    ,energia_percentual6,energia_percentual7,energia_percentual8,energia_percentual9,energia_percentual10)

def calcula_energia_bandas_consecutivas(sinal,fs,janela,passo):
    
    sinal = sinal - np.mean(sinal)
    janela = int(fs * janela) 
    passo = int(fs * passo)
     
    init = 0
    BANDA1 = []
    BANDA2 = []
    BANDA3 = []
    BANDA4 = []
    BANDA5 = []
    BANDA6 = []
    BANDA7 = []
    BANDA8 = []
    BANDA9 = []
    BANDA10 = []
    
    while True:   
        amostras = sinal[init:(init + janela)]
        if(len(amostras) == janela):
            
            banda1,banda2,banda3,banda4,banda5,banda6,banda7,banda8,banda9,banda10 =\
            energias_em_bandas_consecutivas(amostras,fs)
            
            BANDA1 = np.append(BANDA1,banda1)
            BANDA2 = np.append(BANDA2,banda2)
            BANDA3 = np.append(BANDA3,banda3)
            BANDA4 = np.append(BANDA4,banda4)
            BANDA5 = np.append(BANDA5,banda5)
            BANDA6 = np.append(BANDA6,banda6)
            BANDA7 = np.append(BANDA7,banda7)
            BANDA8 = np.append(BANDA8,banda8)
            BANDA9 = np.append(BANDA9,banda9)
            BANDA10 = np.append(BANDA10,banda10)
            
            init += passo
    
        else:
            break        
    
    return BANDA1,BANDA2,BANDA3,BANDA4,BANDA5,BANDA6,BANDA7,BANDA8,BANDA9,BANDA10
    
    
    

## Calculando energia em bandas consecutivas para os diferentes grupos

In [435]:
calcula_energia_bandas_consecutivas(xa, fsa,5,2.5)

(91.01, 11.52, 3.69, 2.09, 0.71, 0.67, 0.27, 0.12, 0.06, 0.36)

In [436]:
calcula_energia_bandas_consecutivas(xb, fsb,5,2.5)

(91.77, 18.86, 3.61, 1.76, 0.57, 0.24, 0.09, 0.04, 0.02, 0.13)

In [437]:
calcula_energia_bandas_consecutivas(xc, fsc,5,2.5)

(59.6, 14.73, 29.29, 4.95, 2.38, 2.25, 0.99, 0.4, 0.29, 0.59)

In [438]:
calcula_energia_bandas_consecutivas(xd, fsd,5,2.5)

(56.56, 15.46, 17.59, 17.68, 5.27, 1.14, 0.31, 0.19, 0.08, 0.12)

In [439]:
calcula_energia_bandas_consecutivas(xe, fse,5,2.5)

(62.55, 16.05, 25.98, 3.48, 1.87, 2.34, 1.13, 0.28, 0.09, 0.27)

## Resultados - (bandas em ordem crescente)


| BANDAS\GRUPO |   A   |   B   |   C   |   D   |   E   |
|--------------|:-----:|:-----:|:-----:|:-----:|:-----:|
|       1      | 91.01 | 91.77 | 59.6  | 56.56 | 62.55 |
|       2      | 11.52 | 18.86 | 14.73 | 15.46 | 16.05 |
|       3      | 3.69  | 3.61  | 29.29 | 17.59 | 25.98 |
|       4      | 2.09  | 1.76  | 4.95  | 17.68 | 3.48  |
|       5      | 0.71  | 0.57  | 2.38  | 5.27  | 1.87  |
|       6      | 0.67  | 0.24  | 2.25  | 1.14  | 2.34  |
|       7      | 0.27  | 0.09  | 0.99  | 0.31  | 1.13  |
|       8      | 0.12  | 0.04  | 0.4   | 0.19  | 0.28  |
|       9      | 0.06  | 0.02  | 0.29  | 0.08  | 0.09  |
|      10      | 0.36  | 0.13  | 0.59  | 0.12  | 0.27  |

## Item d - 


Para a primeira tabela é possível diferenciar o grupo A dos demais por algumas características como:
ter valor RMS inferior ao de B, valor RMS e frequência média inferior ao grupo C, valor RMS,VMR e frequência média inferior ao grupo D e por fim o grupo E apresenta valor RMS e frequência média superiores.
    Na segunda tabela as diferenças entre o grupo A e os demais são:
O grupo A possui menor energia na banda teta que o grupo B, possui maior energia na banda delta e menor energia nas bandas alfa e beta quando comparado aos grupo C,D e E.
    Como há uma semelhança nos dados das duas últimas tabelas, algumas caraterísticas serão semelhantes, mas fica claro que:
    O grupo A possui mais energia nas bandas de maior frequência que B, e possui maior energia na banda 1 e menor na banda 3 que os grupos C,D e E.
    
  

## Item e -

In [79]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/a')
RMS_A = [] 
VMR_A = []
FMED_A = []
FMOD_A = []
FMEDIAN_A = []
DELTA_A = []
TETA_A = []
ALFA_A = []
BETA_A = []
GAMA_A = []
BANDA1_A = []
BANDA2_A = []
BANDA3_A = []
BANDA4_A = []
BANDA5_A = []
BANDA6_A = []
BANDA7_A = []
BANDA8_A = []
BANDA9_A = []
BANDA10_A = []

for i in range(1,101): 
    data = scipy.io.loadmat("x"+str(i)+".mat")
    y = data['x']
    x = [row[0] for row in y]
    fs_mat = data['fs']
    fs = fs_mat[0][0]
    
    RMS_,VMR_,FMED_,FMOD_,FMEDIAN_ = parametros(x,5,2.5,fs,win = None)
    RMS_A = np.append(RMS_A, RMS_) 
    VMR_A = np.append(VMR_A, VMR_)
    FMED_A = np.append(FMED_A, FMED_)
    FMOD_A = np.append(FMOD_A, FMOD_)
    FMEDIAN_A = np.append(FMEDIAN_A, FMEDIAN_)
    
    DELTA_,TETA_,ALFA_,BETA_,GAMA_ = calcula_energia(x,fs,5,2.5)
    DELTA_A = np.append(DELTA_A, DELTA_)
    TETA_A = np.append(TETA_A, TETA_)
    ALFA_A = np.append(ALFA_A, ALFA_)
    BETA_A = np.append(BETA_A, BETA_)
    GAMA_A = np.append(GAMA_A, GAMA_)
    
    BANDA1_,BANDA2_,BANDA3_,BANDA4_,BANDA5_,BANDA6_,BANDA7_,BANDA8_,BANDA9_,BANDA10_ = \
    calcula_energia_bandas_consecutivas(x,fs,5,2.5)
        
    BANDA1_A = np.append(BANDA1_A,BANDA1_)
    BANDA2_A = np.append(BANDA2_A,BANDA2_)
    BANDA3_A = np.append(BANDA3_A,BANDA3_)
    BANDA4_A = np.append(BANDA4_A,BANDA4_)
    BANDA5_A = np.append(BANDA5_A,BANDA5_)
    BANDA6_A = np.append(BANDA6_A,BANDA6_)
    BANDA7_A = np.append(BANDA7_A,BANDA7_)
    BANDA8_A = np.append(BANDA8_A,BANDA8_)
    BANDA9_A = np.append(BANDA9_A,BANDA9_)
    BANDA10_A = np.append(BANDA10_A,BANDA10_)
    
    


In [80]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/b')
RMS_B = [] 
VMR_B = []
FMED_B = []
FMOD_B = []
FMEDIAN_B = []
DELTA_B = []
TETA_B = []
ALFA_B = []
BETA_B = []
GAMA_B = []
BANDA1_B = []
BANDA2_B = []
BANDA3_B = []
BANDA4_B = []
BANDA5_B = []
BANDA6_B = []
BANDA7_B = []
BANDA8_B = []
BANDA9_B = []
BANDA10_B = []

for i in range(101,201): 
    data = scipy.io.loadmat("x"+str(i)+".mat")
    y = data['x']
    x = [row[0] for row in y]
    fs_mat = data['fs']
    fs = fs_mat[0][0]
    
    RMS_,VMR_,FMED_,FMOD_,FMEDIAN_ = parametros(x,5,2.5,fs,win = None)
    RMS_B = np.append(RMS_B, RMS_) 
    VMR_B = np.append(VMR_B, VMR_)
    FMED_B = np.append(FMED_B, FMED_)
    FMOD_B = np.append(FMOD_B, FMOD_)
    FMEDIAN_B = np.append(FMEDIAN_B, FMEDIAN_)
    
    DELTA_,TETA_,ALFA_,BETA_,GAMA_ = calcula_energia(x,fs,5,2.5)
    DELTA_B = np.append(DELTA_B, DELTA_)
    TETA_B = np.append(TETA_B, TETA_)
    ALFA_B = np.append(ALFA_B, ALFA_)
    BETA_B = np.append(BETA_B, BETA_)
    GAMA_B = np.append(GAMA_B, GAMA_)
    
    BANDA1_,BANDA2_,BANDA3_,BANDA4_,BANDA5_,BANDA6_,BANDA7_,BANDA8_,BANDA9_,BANDA10_ = \
    calcula_energia_bandas_consecutivas(x,fs,5,2.5)
        
    BANDA1_B = np.append(BANDA1_B,BANDA1_)
    BANDA2_B = np.append(BANDA2_B,BANDA2_)
    BANDA3_B = np.append(BANDA3_B,BANDA3_)
    BANDA4_B = np.append(BANDA4_B,BANDA4_)
    BANDA5_B = np.append(BANDA5_B,BANDA5_)
    BANDA6_B = np.append(BANDA6_B,BANDA6_)
    BANDA7_B = np.append(BANDA7_B,BANDA7_)
    BANDA8_B = np.append(BANDA8_B,BANDA8_)
    BANDA9_B = np.append(BANDA9_B,BANDA9_)
    BANDA10_B = np.append(BANDA10_B,BANDA10_)

In [81]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/c')
RMS_C = [] 
VMR_C = []
FMED_C = []
FMOD_C = []
FMEDIAN_C = []
DELTA_C = []
TETA_C = []
ALFA_C = []
BETA_C = []
GAMA_C = []
BANDA1_C = []
BANDA2_C = []
BANDA3_C = []
BANDA4_C = []
BANDA5_C = []
BANDA6_C = []
BANDA7_C = []
BANDA8_C = []
BANDA9_C = []
BANDA10_C = []

for i in range(201,301): 
    data = scipy.io.loadmat("x"+str(i)+".mat")
    y = data['x']
    x = [row[0] for row in y]
    fs_mat = data['fs']
    fs = fs_mat[0][0]
    
    RMS_,VMR_,FMED_,FMOD_,FMEDIAN_ = parametros(x,5,2.5,fs,win = None)
    RMS_C = np.append(RMS_C, RMS_) 
    VMR_C = np.append(VMR_C, VMR_)
    FMED_C = np.append(FMED_C, FMED_)
    FMOD_C = np.append(FMOD_C, FMOD_)
    FMEDIAN_C = np.append(FMEDIAN_C, FMEDIAN_)
    
    DELTA_,TETA_,ALFA_,BETA_,GAMA_ = calcula_energia(x,fs,5,2.5)
    DELTA_C = np.append(DELTA_C, DELTA_)
    TETA_C = np.append(TETA_C, TETA_)
    ALFA_C = np.append(ALFA_C, ALFA_)
    BETA_C = np.append(BETA_C, BETA_)
    GAMA_C = np.append(GAMA_C, GAMA_)
    
    BANDA1_,BANDA2_,BANDA3_,BANDA4_,BANDA5_,BANDA6_,BANDA7_,BANDA8_,BANDA9_,BANDA10_ = \
    calcula_energia_bandas_consecutivas(x,fs,5,2.5)
        
    BANDA1_C = np.append(BANDA1_C,BANDA1_)
    BANDA2_C = np.append(BANDA2_C,BANDA2_)
    BANDA3_C = np.append(BANDA3_C,BANDA3_)
    BANDA4_C = np.append(BANDA4_C,BANDA4_)
    BANDA5_C = np.append(BANDA5_C,BANDA5_)
    BANDA6_C = np.append(BANDA6_C,BANDA6_)
    BANDA7_C = np.append(BANDA7_C,BANDA7_)
    BANDA8_C = np.append(BANDA8_C,BANDA8_)
    BANDA9_C = np.append(BANDA9_C,BANDA9_)
    BANDA10_C = np.append(BANDA10_C,BANDA10_)

In [82]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/d')
RMS_D = [] 
VMR_D = []
FMED_D = []
FMOD_D = []
FMEDIAN_D = []
DELTA_D = []
TETA_D = []
ALFA_D = []
BETA_D = []
GAMA_D = []
BANDA1_D = []
BANDA2_D = []
BANDA3_D = []
BANDA4_D = []
BANDA5_D = []
BANDA6_D = []
BANDA7_D = []
BANDA8_D = []
BANDA9_D = []
BANDA10_D = []

for i in range(301,401): 
    data = scipy.io.loadmat("x"+str(i)+".mat")
    y = data['x']
    x = [row[0] for row in y]
    fs_mat = data['fs']
    fs = fs_mat[0][0]
    
    RMS_,VMR_,FMED_,FMOD_,FMEDIAN_ = parametros(x,5,2.5,fs,win = None)
    RMS_D = np.append(RMS_D, RMS_) 
    VMR_D = np.append(VMR_D, VMR_)
    FMED_D = np.append(FMED_D, FMED_)
    FMOD_D = np.append(FMOD_D, FMOD_)
    FMEDIAN_D = np.append(FMEDIAN_D, FMEDIAN_)
    
    DELTA_,TETA_,ALFA_,BETA_,GAMA_ = calcula_energia(x,fs,5,2.5)
    DELTA_D = np.append(DELTA_D, DELTA_)
    TETA_D = np.append(TETA_D, TETA_)
    ALFA_D = np.append(ALFA_D, ALFA_)
    BETA_D = np.append(BETA_D, BETA_)
    GAMA_D = np.append(GAMA_D, GAMA_)
    
    BANDA1_,BANDA2_,BANDA3_,BANDA4_,BANDA5_,BANDA6_,BANDA7_,BANDA8_,BANDA9_,BANDA10_ = \
    calcula_energia_bandas_consecutivas(x,fs,5,2.5)
        
    BANDA1_D = np.append(BANDA1_D,BANDA1_)
    BANDA2_D = np.append(BANDA2_D,BANDA2_)
    BANDA3_D = np.append(BANDA3_D,BANDA3_)
    BANDA4_D = np.append(BANDA4_D,BANDA4_)
    BANDA5_D = np.append(BANDA5_D,BANDA5_)
    BANDA6_D = np.append(BANDA6_D,BANDA6_)
    BANDA7_D = np.append(BANDA7_D,BANDA7_)
    BANDA8_D = np.append(BANDA8_D,BANDA8_)
    BANDA9_D = np.append(BANDA9_D,BANDA9_)
    BANDA10_D = np.append(BANDA10_D,BANDA10_)

In [83]:
os.chdir('/home/gabriel/Documentos/PSB_2018/eeg/e')
RMS_E = [] 
VMR_E = []
FMED_E = []
FMOD_E = []
FMEDIAN_E = []
DELTA_E = []
TETA_E = []
ALFA_E = []
BETA_E = []
GAMA_E = []
BANDA1_E = []
BANDA2_E = []
BANDA3_E = []
BANDA4_E = []
BANDA5_E = []
BANDA6_E = []
BANDA7_E = []
BANDA8_E = []
BANDA9_E = []
BANDA10_E = []

for i in range(401,501): 
    data = scipy.io.loadmat("x"+str(i)+".mat")
    y = data['x']
    x = [row[0] for row in y]
    fs_mat = data['fs']
    fs = fs_mat[0][0]
    
    RMS_,VMR_,FMED_,FMOD_,FMEDIAN_ = parametros(x,5,2.5,fs,win = None)
    RMS_E = np.append(RMS_E, RMS_) 
    VMR_E = np.append(VMR_E, VMR_)
    FMED_E = np.append(FMED_E, FMED_)
    FMOD_E = np.append(FMOD_E, FMOD_)
    FMEDIAN_E = np.append(FMEDIAN_E, FMEDIAN_)
    
    DELTA_,TETA_,ALFA_,BETA_,GAMA_ = calcula_energia(x,fs,5,2.5)
    DELTA_E = np.append(DELTA_E, DELTA_)
    TETA_E = np.append(TETA_E, TETA_)
    ALFA_E = np.append(ALFA_E, ALFA_)
    BETA_E = np.append(BETA_E, BETA_)
    GAMA_E = np.append(GAMA_E, GAMA_)
    
    BANDA1_,BANDA2_,BANDA3_,BANDA4_,BANDA5_,BANDA6_,BANDA7_,BANDA8_,BANDA9_,BANDA10_ = \
    calcula_energia_bandas_consecutivas(x,fs,5,2.5)
        
    BANDA1_E = np.append(BANDA1_E,BANDA1_)
    BANDA2_E = np.append(BANDA2_E,BANDA2_)
    BANDA3_E = np.append(BANDA3_E,BANDA3_)
    BANDA4_E = np.append(BANDA4_E,BANDA4_)
    BANDA5_E = np.append(BANDA5_E,BANDA5_)
    BANDA6_E = np.append(BANDA6_E,BANDA6_)
    BANDA7_E = np.append(BANDA7_E,BANDA7_)
    BANDA8_E = np.append(BANDA8_E,BANDA8_)
    BANDA9_E = np.append(BANDA9_E,BANDA9_)
    BANDA10_E = np.append(BANDA10_E,BANDA10_)

# 1 - Identificar participantes que não apresentam quadro de epilepsia

## Hipótese alternativa 1: sinais do grupo A possuem mais energia na BANDA1 que os sinais C,D e E

### Teste de normalidade dos sinais

In [94]:
p,a = scipy.stats.shapiro(BANDA1_A)
p

0.9384085536003113

### Considerando um alpha de 0.05, o teste não rejeitou a hipótese nula, dessa forma utilizar um teste não paramétrico

In [96]:
p = scipy.stats.ranksums(BANDA1_A, BANDA1_C)
p

RanksumsResult(statistic=34.525439562942076, pvalue=3.331410576821401e-261)

In [97]:
p = scipy.stats.ranksums(BANDA1_A, BANDA1_D)
p

RanksumsResult(statistic=28.53647183916271, pvalue=4.1347238435079335e-179)

In [98]:
p = scipy.stats.ranksums(BANDA1_A, BANDA1_E)
p

RanksumsResult(statistic=32.61266233597111, pvalue=2.7129167589212394e-233)

### Os testes rejeitaram a hipótese nula

## Hipótese alternativa 2: Frequência média dos grupos C,D e E superiores ao grupo A

In [101]:
p,a = scipy.stats.shapiro(FMED_A)
p

0.9855228066444397

In [102]:
p,a = scipy.stats.shapiro(FMED_C)
p

0.9612719416618347

In [103]:
p,a = scipy.stats.shapiro(FMED_D)
p

0.9700703024864197

In [104]:
p,a = scipy.stats.shapiro(FMED_E)
p

0.9767362475395203

## Considerando um alfa de 0.05, nenhum dos testes rejeitou a hipótese nula de que os dados foram extraídos de uma distribuição normal. Utilzando então um teste paramétrico (t-Student)

In [107]:
t,p = scipy.stats.ttest_ind(FMED_A,FMED_C)
p

0.0

In [109]:
t,p = scipy.stats.ttest_ind(FMED_A,FMED_D)
p

8.689732785219432e-222

In [110]:
t,p = scipy.stats.ttest_ind(FMED_A,FMED_E)
p

3.801874019059368e-240

# Resultado - Considerando um alfa de 0.05 a hipótese nula foi rejeitada em todos os testes.

# 2 - Identificar participantes onde os sinais foram adquiridos com os olhos abertos.

### Hipótese alternativa 1 - Os participantes do grupo A possuem no sinal de EEG maior concentração de energia na banda Delta

### teste de normalidade


In [113]:
p,a = scipy.stats.shapiro(DELTA_A)
p

0.9724797606468201

In [114]:
p,a = scipy.stats.shapiro(DELTA_C)
p

0.9312717914581299

### Considerando um alpha de 0.05, o teste para a amostra DELTA_C não rejeitou a hipótese nula, dessa forma utilizar um teste não paramétrico

In [115]:
p = scipy.stats.ranksums(DELTA_A, DELTA_C)
p

RanksumsResult(statistic=32.459748377175636, pvalue=3.946326358625924e-231)

## Hipótese alternativa 2 - Os participantes do grupo A possui maior energia na segunda banda consecutiva comparado ao grupo B

### Teste de normalidade

In [116]:
p,a = scipy.stats.shapiro(BANDA2_A)
p

0.9230820536613464

In [117]:
p,a = scipy.stats.shapiro(BANDA2_B)
p

0.9420458078384399

### Considerando um alpha de 0.05, o teste para as amostra  não rejeitou a hipótese nula, dessa forma utilizar um teste não paramétrico

In [118]:
p = scipy.stats.ranksums(BANDA2_A, BANDA2_B)
p

RanksumsResult(statistic=6.216444823133985, pvalue=5.085450909432012e-10)

# Resultado - Considerando um alfa de 0.05 a hipótese nula foi rejeitada.

# 3 - Identificar participantes onde os sinais foram adquiridos na zona epileptogênica.


### Hipótese alternativa 1 - Os participantes do grupo D possuem maior valor RMS que os participantes do grupo C

### Teste de normalidade

In [120]:
p,a = scipy.stats.shapiro(RMS_D)
p

0.9425197243690491

In [121]:
p,a = scipy.stats.shapiro(RMS_C)
p

0.9343847632408142

### Considerando um alpha de 0.05, o teste para as amostra  não rejeitou a hipótese nula, dessa forma utilizar um teste não paramétrico

In [122]:
p = scipy.stats.ranksums(RMS_D, RMS_C)
p

RanksumsResult(statistic=34.421332536501, pvalue=1.2094457584343503e-259)

# Resultado - Considerando um alfa de 0.05 a hipótese nula foi rejeitada.

# 4 -  Identificar entre aqueles que possuem quadro de epilepsia, os sinais proveniente do grupo em que os sinais foram adquiridos durante convulsões

## Hipótese alternativa 1 - Os participantes do grupo E possui maior energia na BANDA 1 que os sinais C e D

### Teste de Normalidade

In [123]:
p,a = scipy.stats.shapiro(BANDA1_E)
p

0.9943946599960327

In [124]:
p,a = scipy.stats.shapiro(BANDA1_C)
p

0.9845269918441772

In [125]:
p,a = scipy.stats.shapiro(BANDA1_D)
p

0.9156437516212463

## Utilizar um teste não paramétrico entre as amostras E e D

In [126]:
t,p = scipy.stats.ttest_ind(BANDA1_E, BANDA1_C)
p

4.499519517209845e-230

In [129]:
p = scipy.stats.ranksums(BANDA1_E, BANDA1_D)
p

RanksumsResult(statistic=-9.036424963468324, pvalue=1.618794025578888e-19)

# Resultado - Considerando um alfa de 0.05 a hipótese nula foi rejeitada.